In [49]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [50]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [51]:
df_market_data.dtypes

price_change_percentage_24h     float64
price_change_percentage_7d      float64
price_change_percentage_14d     float64
price_change_percentage_30d     float64
price_change_percentage_60d     float64
price_change_percentage_200d    float64
price_change_percentage_1y      float64
dtype: object

In [52]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [53]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [54]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
# Scaling the numeric columns
market_data_scalled=StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d","price_change_percentage_30d","price_change_percentage_60d","price_change_percentage_200d","price_change_percentage_1y"]])

In [81]:
# Create a DataFrame with the scaled data, 
# Copy the crypto names from the original data
# Set the coinid column as index

df_market_transformed=pd.DataFrame(
    market_data_scalled, 
    columns=["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d","price_change_percentage_30d","price_change_percentage_60d","price_change_percentage_200d","price_change_percentage_1y"],
    index=df_market_data.index
    )

# Display sample data

df_market_transformed.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Data.

In [56]:
# Create a list with the number of k-values from 1 to 11
k_values =list(range(1,11))

In [57]:
# Create an empty list to store the inertia values
inertia_values=[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k_values:
    k_model=KMeans(n_clusters=i, random_state=1)
    k_model.fit(df_market_transformed)
    inertia_values.append(k_model.inertia_)


In [58]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data={"k":k_values, "inertia":inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
# Review the DataFrame
df_elbow.head(5)

k     inertia
0  1  287.000000
1  2  212.123342
2  3  165.136752
3  4   79.022435
4  5   66.413051

In [59]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
lock_elbow_plot = df_elbow.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve',
    xticks=k_values,
)

lock_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [60]:
# Initialize the K-Means model using the best value for k
model =KMeans(n_clusters=4, random_state=1)

In [61]:
# Fit the K-Means model using the scaled data
model.fit(df_market_transformed)

KMeans(n_clusters=4, random_state=1)

In [62]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_2= model.predict(df_market_transformed)

# Print the resulting array of cluster values.
k_2

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0],
      dtype=int32)

In [63]:
# Create a copy of the DataFrame
df_model=df_market_transformed.copy()

In [64]:
# Add a new column to the DataFrame with the predicted clusters
df_model["Cluster Predictions"]=k_2

# Display sample data
df_model.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  Cluster Predictions  
coin_id                                                           
bitcoin                           -0.251637                    2  
ethereum                          -0.199352                    2  
tether                            -0.282061                    0  
ripple                            -0.295546                    0  
bitcoin-cash                      -0.270317                    2  
binancecoin                       -0.225533                    2  
chainlink                         -0.018284                    2  
cardano                           -0.155428                    2  
litecoin                          -0.292351                    2  
bitcoin-cash-sv                   -0.206029                    0

In [65]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
cluster_01=df_model.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Cluster Predictions",
    hover_cols=["coin_id"]
)
cluster_01

:NdOverlay   [Cluster Predictions]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [66]:
# Create a PCA model instance and set `n_components=3`.
pca_val=PCA(n_components=3)
df_market_pca=df_model.drop(columns=["Cluster Predictions"])

In [67]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
transformed_data = pca_val.fit_transform(df_market_transformed)

# View the first five rows of the DataFrame. 
print(transformed_data[:5])

[[-0.60066733  0.84276006  0.46159457]
 [-0.45826071  0.45846566  0.95287678]
 [-0.43306981 -0.16812638 -0.64175193]
 [-0.47183495 -0.22266008 -0.47905316]
 [-1.15779997  2.04120919  1.85971527]]


In [68]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
print(f"Explained Variance: {pca_val.explained_variance_}")

# Total variance explained by the three principal components
print(f"Total Variance Explained: {sum(pca_val.explained_variance_ratio_)}")



Explained Variance: [2.6689967  2.48978331 1.26307212]
Total Variance Explained: 0.8950316570309842


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8950316570309842

In [69]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
# Copy the crypto names from the original data
# Set the coinid column as index

# Display sample data
df_market_pca=pd.DataFrame(
    transformed_data,
    columns=["PC1", "PC2", "PC3"],
    index=df_market_data.index

)
df_market_pca.head(10)

PC1       PC2       PC3
coin_id                                      
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the PCA Data

In [70]:
# Create a list with the number of k-values from 1 to 11
k_2020=list(range(1,11))

In [71]:
# Create an empty list to store the inertia values
inertia_values_2020=[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k_2020:
    k_model=KMeans(n_clusters=i, random_state=1)
    k_model.fit(df_market_pca)
    inertia_values_2020.append(k_model.inertia_)


In [72]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_2020={"k":k_2020, "inertia":inertia_values_2020}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_2020 = pd.DataFrame(elbow_data_2020)

# Review the DataFrame
df_elbow_2020.head(5)

k     inertia
0  1  256.874086
1  2  182.339530
2  3  135.442408
3  4   49.665497
4  5   38.672582

In [73]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
loose_elbow_plot = df_elbow_2020.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve',
    xticks=k_2020,
)

loose_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [74]:
# Initialize the K-Means model using the best value for k
model_2020 =KMeans(n_clusters=4, random_state=1)

In [75]:
# Fit the K-Means model using the PCA data
model_2020.fit(df_market_pca)

KMeans(n_clusters=4, random_state=1)

In [76]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4= model_2020.predict(df_market_pca)
# Print the resulting array of cluster values.
k_4

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0],
      dtype=int32)

In [77]:
# Create a copy of the DataFrame with the PCA data
k4_model_pca=df_market_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
k4_model_pca["Cluster Predictions"]=k_4

# Display sample data
k4_model_pca.head(10)

PC1       PC2       PC3  Cluster Predictions
coin_id                                                           
bitcoin         -0.600667  0.842760  0.461595                    2
ethereum        -0.458261  0.458466  0.952877                    2
tether          -0.433070 -0.168126 -0.641752                    0
ripple          -0.471835 -0.222660 -0.479053                    0
bitcoin-cash    -1.157800  2.041209  1.859715                    2
binancecoin     -0.516534  1.388377  0.804071                    2
chainlink       -0.450711  0.517699  2.846143                    2
cardano         -0.345600  0.729439  1.478013                    2
litecoin        -0.649468  0.432165  0.600303                    2
bitcoin-cash-sv -0.759014 -0.201200 -0.217653                    0

In [78]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
cluster_04 = cluster_04=k4_model_pca.hvplot.scatter(
    x='PC1',
    y='PC2',
    by='Cluster Predictions',
    hover_cols=['coin_id']
)

cluster_04

:NdOverlay   [Cluster Predictions]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [79]:
# Composite plot to contrast the Elbow curves
cluster_analysis_elbow_curve = lock_elbow_plot * loose_elbow_plot
cluster_analysis_elbow_curve



:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [80]:
# Composite plot to contrast the clusters
cluster_analysis = cluster_01 * cluster_04
cluster_analysis

:Overlay
   .NdOverlay.I  :NdOverlay   [Cluster Predictions]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [Cluster Predictions]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The elbow method curve graphically displays comparable slopes and noticeable bends. The PCA-derived cluster plots exhibit tighter clusters, indicating greater accuracy in clustering.